src/viztracer/decorator.py

```diff
@@ -57,47 +57,83 @@ def wrapper(*args, **kwargs) -> Any:
     return inner
 
 
-def log_sparse(func: Optional[Callable] = None, stack_depth: int = 0) -> Callable:
-    tracer = get_tracer()
-    if tracer is None or not tracer.log_sparse:
-        if func is None:
-            return lambda f: f
-        return func
+def _log_sparse_wrapper(func: Optional[Callable], stack_depth: int = 0,
+                        dynamic_tracer_check: bool = False) -> Callable:
+    if not dynamic_tracer_check:
+        tracer = get_tracer()
+        if tracer is None or not tracer.log_sparse:
+            return func
+
+    if stack_depth > 0:
```

Comment:
Wow this part is getting out of hand fast :)

So here's my suggestion:
1. We used to have two wrappers depending on whether `func` is `None - that's because the layers are different. Now that we are using `partial`, we can use just one `wrapper` and check `stack_depth` when we need to log the trace. Yes it would introduce an extra run-time if statement, but the overhead is already large in this case. We could reduce plenty of redundant code.
2. The if logic below is getting crazy - even I would have a very difficult time following it. We should try to sort out the logic and simplify that. When we are on the if statement, we basically only have two options:
    * Execute the function normally but make sure VizTracer does not ignore it
    * Do a manual sparse log, but two methods based on `stack_depth`
3. We should only do the latter when `log_sparse==True` and `tracer.enable==False`, for all other cases, we should do the former.
4. It's okay to have a private method of tracer to `setignorestackcounter`, but can we do a public method of VizTracer like `shield_ignore()` which is a context so we can do
    ```python
    with tracer.shield_ignore():
        func()
    ```
    This requires some count manipulation for the context method (because the counter would be changing going inside and outside of the method), but it would be a one time thing and we can use that later.
5. `shield_ignore()` should be a no-op when the tracer is disabled (stopped) or when the tracer is not in ignore state (I think it should just work if the logic is implemented correctly), so we do not even need to have a separate case for "normal execution" (your current `else` case).

In this way, I believe we can greatly reduce the complexity (both length and logic) of the decorator and make this much easier to read and understand.